core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -4764,6 +4764,123 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
   def serviceWithZIO[Service]: ServiceWithZIOPartiallyApplied[Service] =
     new ServiceWithZIOPartiallyApplied[Service]
 
+  /**
+   * A `ZIOFunctionConstructor[Input]` knows how to construct a `ZIO` value from
+   * a function of type `In`. This allows the type of the `ZIO` value
+   * constructed to depend on `In`.
+   */
+  trait ZIOFunctionConstructor[In] {
+
+    /**
+     * The type of the `ZIO` value.
+     */
+    type Out
+
+    /**
+     * Constructs a `ZIO` value from the specified input.
+     */
+    def apply(in: In)(implicit trace: Trace): Out
+  }
+
+  object ZIOFunctionConstructor {
+
+    type WithOut[In, Out0] = ZIOFunctionConstructor[In] { type Out = Out0 }
+
+    implicit def function1Constructor[A: Tag, Z: Tag, R, E]
+      : ZIOFunctionConstructor.WithOut[A => ZIO[R, E, Z], ZIO[R with A, E, Z]] =
+      new ZIOFunctionConstructor[A => ZIO[R, E, Z]] {
+        type Out = ZIO[R with A, E, Z]
+        def apply(f: A => ZIO[R, E, Z])(implicit trace: Trace): ZIO[R with A, E, Z] =
+          ZIO.serviceWith[A](a => f(a)).flatten
+
+      }
+
+    implicit def function2Constructor[A: Tag, B: Tag, Z: Tag, R, E]
+      : ZIOFunctionConstructor.WithOut[(A, B) => ZIO[R, E, Z], ZIO[R with A with B, E, Z]] =
+      new ZIOFunctionConstructor[(A, B) => ZIO[R, E, Z]] {
+        type Out = ZIO[R with A with B, E, Z]
+        def apply(f: (A, B) => ZIO[R, E, Z])(implicit trace: Trace): ZIO[R with A with B, E, Z] =
+          for {
+            a <- ZIO.service[A]
+            b <- ZIO.service[B]
+            z <- f(a, b)
+          } yield z
```

Comment:
This can be done in a single suspension:

```suggestion
            ZIO.environmentWith[A with B] { env =>
            	f(a, b)
            }
```

Goes for all the other variants 🙂

Reply:
Yes please, this way we avoid all the unnecessary flatmaps

Just a minor correction, I think the code would be:
```scala
f(env.get[A], env.get[B])
```

Reply:
Yep, thanks

Reply:
Done for all